gk/main.c

```diff
@@ -548,6 +555,67 @@ gk_setup_rss(struct gk_config *gk_conf)
 	return ret;
 }
 
+static void
+gk_arp_and_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	struct ether_addr *nexthop_mac = arg;
+
+	/* TODO Add concurrency control by sequential lock on nexthop_mac. */
+	if (!map->stale)
+		ether_addr_copy(&map->ha, nexthop_mac);
+	else
+		/* 00:00:00:00:00:00 represents invalid MAC address. */
+		memset(nexthop_mac, 0, sizeof(*nexthop_mac));
+
+	*pcall_again = true;
+}
+
+static int
+gk_hold_arp_and_nd(struct gk_lpm *ltbl, unsigned int lcore_id)
+{
+	unsigned int i;
+	int ret;
+	struct gk_policy *policies = ltbl->policy_tbl;
+	for (i = 0; i < ltbl->num_policies; i++) {
+		struct ipip_tunnel_info *tunnel;
+
+		if (policies[i].action != GK_FWD_GT)
+			continue;
+
+		tunnel = &policies[i].grantor_tunnel;
+		if (tunnel->flow.proto == ETHER_TYPE_IPv4) {
+			struct in_addr *ipv4_dst;
+			if (policies[i].is_gt_in_back_net)
+				ipv4_dst = (struct in_addr *)
+					&tunnel->flow.f.v4.dst;
+			else
+				ipv4_dst = &policies[i].gw->ip.v4;
+
+			ret = hold_arp(gk_arp_and_nd_req_cb, (void *)
```

Comment:
(1) That makes sense. So, I can remove the lines of code that copying the Ethernet header from ``encapsulate()``.

(2) For the current implementation, to minimize the number of policies, I stored all the gateway entries in the array ``struct gk_gateway gateways[GK_MAX_NUM_GW]``, instead of adding a separate policy for each of the gateways. So, it won't conduct the second lookup of the LPM table. For the action ``GK_FWD_GATEWAY``, the LLS will update the MAC of the Ethernet header referenced by ``struct gk_gateway *gw``. So, it should always obtain the latest MAC address. However, by using two LPM lookups for GK_FWD_GATEWAY, it may lead to more policies.